Separated resource rendering into another trait

teinarss opened the pull request:
Resurrection of #13902

Added the ResourceRenderer  to the player traits folder cuz of the comment from pchote in the previous PR

teinarss committed:
Separated resource rendering into another trait

reaperrr commented on mods/ts/rules/world.yaml:
```diff
@@ -279,6 +279,8 @@ World:
 		Sequence: largecraters
 	ResourceLayer:
+	ResourceRenderer:
+		RenderTypes: Tiberium, BlueTiberium
```
What about Veins?

teinarss commented:
Updated: Fixed veins

abcdefg30 commented on OpenRA.Mods.D2k/Traits/World/D2kResourceRenderer.cs:
```diff
@@ -139,46 +139,63 @@ ClearSides FindClearSides(ResourceType t, CPos p)
+				}
 				else
-					t.Sprite = null;
+					throw new InvalidOperationException("Couldn't find a suitable sprite??");	// TODO: Handle this better?
```
What is the plan here? We could fallback to rendering empty content (but I think throwing is better than having bogus state). If we are 100% sure that this branch will never be taken we could remove it if we don't want any saveguards. Tl;dr: What about `throw new InvalidOperationException("SpriteMap does not contain an index for ClearSides type '{0}'".F(clear));`?

abcdefg30 commented on OpenRA.Mods.Common/Traits/Player/ResourceRenderer.cs:
```diff
@@ -0,0 +1,196 @@
+
+		readonly HashSet<CPos> dirty = new HashSet<CPos>();
+
+		readonly Dictionary<PaletteReference, TerrainSpriteLayer> spriteLayers = new Dictionary<PaletteReference, TerrainSpriteLayer>();
```
Imho it would be better to have the `protected` and not `protected` as two groups without newlines between the group members.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Player/ResourceRenderer.cs:
```diff
@@ -0,0 +1,196 @@
+	public class ResourceRendererInfo : ITraitInfo, Requires<ResourceLayerInfo>
+	{
+		[FieldLoader.Require]
+		[Desc("Only render these ResourceType Type names.")]
```
`Only render these ResourceType names.`? (Or lowercase `type`, but imho that sentence would then still be odd.)

abcdefg30 reviewed (commented):
Code looks otherwise good on a first pass.

This will want an update rule.

teinarss commented:
Updated

abcdefg30 commented on OpenRA.Mods.Common/UpdateRules/Rules/AddResourceRenderer.cs:
```diff
@@ -0,0 +1,65 @@
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			if (!actorNode.ChildrenMatching("ResourceLayer").Any() && !actorNode.ChildrenMatching("ResourceRenderer").Any())
```
Doesn't this have to be `actorNode.ChildrenMatching("ResourceLayer").Any() && !actorNode.ChildrenMatching("ResourceRenderer").Any()`?

abcdefg30 commented on OpenRA.Mods.Common/UpdateRules/Rules/AddResourceRenderer.cs:
```diff
@@ -0,0 +1,65 @@
+				actorNode.AddNode(resourceRenderer);
+			}
+
+			if (!actorNode.ChildrenMatching("D2kResourceLayer").Any() && !actorNode.ChildrenMatching("D2kResourceRenderer").Any())
```
Same here.

abcdefg30 commented on OpenRA.Mods.D2k/Traits/World/D2kResourceRenderer.cs:
```diff
@@ -101,12 +100,13 @@ public enum ClearSides : byte
 		{
-			return RenderContent.Contains(c) && RenderContent[c].Type == t;
+			var type = ResourceLayer.GetResourceType(c);
+			return RenderContent.Contains(c) && type == t;
```
(Very) Minor nit: If `!RenderContent.Contains(c)` we'd unnecessarily call `GetResourceType`, maybe `return RenderContent.Contains(c) && ResourceLayer.GetResourceType(c) == t;` is better.

abcdefg30 reviewed (commented):
This doesn't look correct. (?)
![grafik](https://user-images.githubusercontent.com/7704140/70471772-2be16b80-1ace-11ea-9a7e-a8df28d192c9.png)

teinarss commented:
> 
> 
> This doesn't look correct. (?)
> ![grafik](https://user-images.githubusercontent.com/7704140/70471772-2be16b80-1ace-11ea-9a7e-a8df28d192c9.png)

What map is this and where on the map?

teinarss commented:
Updated

abcdefg30 commented:
> What map is this and where on the map?

Battle for Dune in the upper right.

teinarss commented:
> 
> 
> > What map is this and where on the map?
> 
> Battle for Dune in the upper right.

Need more info, still cant reproduce it. I did rebase it on bleed. Is in still present for you?

abcdefg30 commented:
Testcase is simply revealing what another player harvested.

teinarss commented:
Updated

teinarss commented:
Made the following improvements too:
* Removed the unnecessary `Sprite`  from `CellContents`. It was only needed because `spriteLayers` was updated from `CellEntryChanged`. Updating the `spriteLayers`  directly now instead of going through the celllayer.
* Removed some unnecessary  access to the CellLayers. Now these values are passed to the needed methods.
* Removed one unnecessary allocation.

pchote reviewed (commented):
Looks mostly good to me. Spotted one regression and have a couple of questions/suggestions.

teinarss commented:
Updated

pchote commented on OpenRA.Mods.Common/Traits/World/ResourceRenderer.cs:
```diff
@@ -0,0 +1,208 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
2020

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/AddResourceRenderer.cs:
```diff
@@ -0,0 +1,65 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
2020

teinarss commented:
Upated

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/48208a712ac54b24cc44f3373fc060b124bee3dc...75f75c281e39f0ea073460e39d51387e3cf2d84d)
